### ISSUE TITLE
Regent: index launch failure with nested partition

### ISSUE THREAD

**Issue by syamajala** [StanfordLegion/legion#1280]
The following example only runs with `__forbid(__index_launch)`:

```
import "regent"
local c = regentlib.c

c.printf.replicable = true

local get_partition_color = regentlib.macro(function(lp)
  return rexpr
    regentlib.c.legion_index_partition_get_color(
      __runtime(), __raw(lp).index_partition)
  end
end)

local import_cross_product = regentlib.macro(function(...)
  local args = terralib.newlist({...})
  local partitions = args:sub(1, #args - 1)
  local colors = args[#args]
  return rexpr
    __import_cross_product(
        partitions, colors,
        regentlib.c.legion_terra_index_cross_product_import(
          __runtime(), __raw([ partitions[1] ]).index_partition))
  end
end)

task init_part(lr_int   : region(ispace(int3d), double),
               color    : int3d,
               subranks : int)
where
  writes(lr_int)
do
  var bounds = lr_int.bounds

  var blocking_factor = ((bounds.hi.z - bounds.lo.z + 1) + subranks - 1) / subranks

  var e = rect3d{ bounds.lo, int3d{bounds.hi.x, bounds.hi.y, bounds.lo.z+blocking_factor-1} }

  var t : transform(3, 1)
  t[{0, 0}] = 0
  t[{1, 0}] = 0
  t[{2, 0}] = blocking_factor

  var is_subrank = ispace(int1d, subranks)
  var lp_subrank = restrict(disjoint, complete, lr_int, t, e, is_subrank)

  c.printf("COLOR: %d %d %d Lo: %d %d %d Hi: %d %d %d Blocking Factor: %d\n",
           color.x, color.y, color.z,
           bounds.lo.x, bounds.lo.y, bounds.lo.z,
           bounds.hi.x, bounds.hi.y, bounds.hi.z,
           blocking_factor)

  for subrank_color in is_subrank do
    var subrank_bounds = lp_subrank[subrank_color].bounds
    c.printf("\tCOLOR: %d %d %d SUBRANK COLOR: %d Lo: %d %d %d Hi: %d %d %d Vol: %d\n",
             color.x, color.y, color.z, subrank_color,
             subrank_bounds.lo.x, subrank_bounds.lo.y, subrank_bounds.lo.z,
             subrank_bounds.hi.x, subrank_bounds.hi.y, subrank_bounds.hi.z,
             lp_subrank[subrank_color].volume)
  end

  return get_partition_color(lp_subrank)
end

__demand(__cuda)
task test(lr_int : region(ispace(int3d), double),
          color : int4d)
  var bounds = lr_int.bounds
  c.printf("COLOR: %d %d %d %d Lo: %d %d %d Hi: %d %d %d\n",
           color.x, color.y, color.z, color.w,
           bounds.lo.x, bounds.lo.y, bounds.lo.z,
           bounds.hi.x, bounds.hi.y, bounds.hi.z)
end

__demand(__replicable)
task main()
  var local_grid_size = int3d{64, 64, 64}
  var proc_grid_size = int3d{1, 1, 2}
  var global_grid_size = local_grid_size * proc_grid_size
  c.printf("GLOBAL GRID SIZE: %d %d %d\n",
           global_grid_size.x, global_grid_size.y, global_grid_size.z)
  var is_grid = ispace(int3d, global_grid_size)

  var blocking_factor = global_grid_size / proc_grid_size
  c.printf("RANK BLOCKING FACTOR: %d %d %d\n",
           blocking_factor.x, blocking_factor.y, blocking_factor.z)
  var ranks = proc_grid_size.x * proc_grid_size.y * proc_grid_size.z
  var subranks = 4
  c.printf("RANKS: %d SUBRANKS: %d\n", ranks, subranks)

  var is_bounds = is_grid.bounds
  var is_rank = ispace(int3d, int3d{
                           (((is_bounds.hi.x - is_bounds.lo.x) + blocking_factor.x) / blocking_factor.x),
                           (((is_bounds.hi.y - is_bounds.lo.y) + blocking_factor.y) / blocking_factor.y),
                           (((is_bounds.hi.z - is_bounds.lo.z) + blocking_factor.z) / blocking_factor.z)},
                         int3d{0, 0, 0})

  var is_subrank = ispace(int1d, subranks)

  var lr_int = region(is_grid, double)
  fill(lr_int, 0)

  var t : transform(3, 3)
  for i = 0, 3 do
    for j = 0, 3 do
      t[{i, j}] = 0
    end
  end

  t[{0, 0}] = blocking_factor.x
  t[{1, 1}] = blocking_factor.y
  t[{2, 2}] = blocking_factor.z

  var e = rect3d{ int3d{0, 0, 0}, blocking_factor - int3d{1, 1, 1}}

  var lp_int_rank = restrict(disjoint, complete, lr_int, t, e, is_rank)

  var rank_color = get_partition_color(lp_int_rank)
  var subrank_color : regentlib.c.legion_color_t = -1

  for color in is_rank do
    subrank_color = init_part(lp_int_rank[color], color, subranks)
  end

  var fake_subrank = partition(equal, lr_int, is_subrank)

  var colors2 : regentlib.c.legion_color_t[2]
  colors2[0] = rank_color
  colors2[1] = subrank_color
  var cp2 = import_cross_product(lp_int_rank, fake_subrank, colors2)

  var rank_subrank = ispace(
    int4d,
    { is_rank.bounds.hi.x - is_rank.bounds.lo.x + 1,
      is_rank.bounds.hi.y - is_rank.bounds.lo.y + 1,
      is_rank.bounds.hi.z - is_rank.bounds.lo.z + 1,
      is_subrank.bounds.hi - is_subrank.bounds.lo + 1 },
    { is_rank.bounds.lo.x, is_rank.bounds.lo.y, is_rank.bounds.lo.z, is_subrank.bounds.lo })

  --__forbid(__index_launch)
  for i in rank_subrank do
    test(cp2[int3d{i.x, i.y, i.z}][i.w], i)
  end
end

regentlib.start(main)
```

When trying to index launch the leaf task I see the following:
```
...j/legion_s3d_test/legion/language/terra/src/terralib.lua:1202: Errors reported during typechecking.
/lustre/scratch/vsyamaj/legion_s3d_test/legion/language/src/regent/std.t:1477: invalid conversion from int1d to int3d
    return `([to]([expr]))
                      ^
/lustre/scratch/vsyamaj/legion_s3d_test/legion/language/src/regent/std.t:1477: user-defined cast failed: ...vsyamaj/legion_s3d_test/legion/language/src/regent/std.t:2385: assertion failed!
stack traceback:
        [C]: in function 'assert'
        ...vsyamaj/legion_s3d_test/legion/language/src/regent/std.t:2385: in function <...vsyamaj/legion_s3d_test/legion/language/src/regent/std.t:2360>
        [C]: in function 'xpcall'
        /lustre/scratch/vsyamaj/legion_s3d_test/legion/language/src/regent/std.t:1477: Errors reported during invoking __cast
    return `([to]([expr]))
                      ^
        ...j/legion_s3d_test/legion/language/terra/src/terralib.lua:2290: in function 'userfn'
        /lustre/scratch/vsyamaj/legion_s3d_test/legion/language/src/regent/std.t:1477: Errors reported during invoking macro
    return `([to]([expr]))
                 ^
        ...j/legion_s3d_test/legion/language/terra/src/terralib.lua:2859: in function 'docheck'
        ...j/legion_s3d_test/legion/language/terra/src/terralib.lua:3120: in function 'checkexp'
        ...j/legion_s3d_test/legion/language/terra/src/terralib.lua:3435: in function 'typecheck'
        ...j/legion_s3d_test/legion/language/terra/src/terralib.lua:1202: in function 'definequote'
        ...vsyamaj/legion_s3d_test/legion/language/src/regent/std.t:1477: in function 'implicit_cast'
        ...maj/legion_s3d_test/legion/language/src/regent/codegen.t:2403: in function 'expr'
        ...maj/legion_s3d_test/legion/language/src/regent/codegen.t:9231: in function 'fn'
        src/terralist.lua:150: in function 'block'
        ...maj/legion_s3d_test/legion/language/src/regent/codegen.t:11172: in function 'compile_thunk'
        ...aj/legion_s3d_test/legion/language/src/regent/std_base.t:1632: in function 'complete'
        ...vsyamaj/legion_s3d_test/legion/language/src/regent/std.t:4518: in function <...vsyamaj/legion_s3d_test/legion/language/src/regent/std.t:4516>
        ...vsyamaj/legion_s3d_test/legion/language/src/regent/std.t:4520: in function 'compile_tasks_in_parallel'
        ...vsyamaj/legion_s3d_test/legion/language/src/regent/std.t:4654: in function 'start'
        test3.rg:144: in main chunk
    return `([to]([expr]))
                      ^
/lustre/scratch/vsyamaj/legion_s3d_test/legion/language/src/regent/std.t:1477: user-defined cast failed: ...vsyamaj/legion_s3d_test/legion/language/src/regent/std.t:2385: assertion failed!
stack traceback:
        [C]: in function 'assert'
        ...vsyamaj/legion_s3d_test/legion/language/src/regent/std.t:2385: in function <...vsyamaj/legion_s3d_test/legion/language/src/regent/std.t:2360>
        [C]: in function 'xpcall'
        /lustre/scratch/vsyamaj/legion_s3d_test/legion/language/src/regent/std.t:1477: Errors reported during invoking __cast
    return `([to]([expr]))
                      ^
        ...j/legion_s3d_test/legion/language/terra/src/terralib.lua:2290: in function 'userfn'
        /lustre/scratch/vsyamaj/legion_s3d_test/legion/language/src/regent/std.t:1477: Errors reported during invoking macro
    return `([to]([expr]))
                 ^
        ...j/legion_s3d_test/legion/language/terra/src/terralib.lua:2859: in function 'docheck'
        ...j/legion_s3d_test/legion/language/terra/src/terralib.lua:3120: in function 'checkexp'
        ...j/legion_s3d_test/legion/language/terra/src/terralib.lua:3435: in function 'typecheck'
        ...j/legion_s3d_test/legion/language/terra/src/terralib.lua:1202: in function 'definequote'
        ...vsyamaj/legion_s3d_test/legion/language/src/regent/std.t:1477: in function 'implicit_cast'
        ...maj/legion_s3d_test/legion/language/src/regent/codegen.t:2403: in function 'expr'
        ...maj/legion_s3d_test/legion/language/src/regent/codegen.t:9231: in function 'fn'
        src/terralist.lua:150: in function 'block'
        ...maj/legion_s3d_test/legion/language/src/regent/codegen.t:11172: in function 'compile_thunk'
        ...aj/legion_s3d_test/legion/language/src/regent/std_base.t:1632: in function 'complete'
        ...vsyamaj/legion_s3d_test/legion/language/src/regent/std.t:4518: in function <...vsyamaj/legion_s3d_test/legion/language/src/regent/std.t:4516>
        ...vsyamaj/legion_s3d_test/legion/language/src/regent/std.t:4520: in function 'compile_tasks_in_parallel'
        ...vsyamaj/legion_s3d_test/legion/language/src/regent/std.t:4654: in function 'start'
        test3.rg:144: in main chunk
    return `([to]([expr]))
                      ^

stack traceback:
        [C]: in function 'error'
        ...j/legion_s3d_test/legion/language/terra/src/terralib.lua:396: in function 'finishandabortiferrors'
        ...j/legion_s3d_test/legion/language/terra/src/terralib.lua:3437: in function 'typecheck'
        ...j/legion_s3d_test/legion/language/terra/src/terralib.lua:1202: in function 'definequote'
        ...vsyamaj/legion_s3d_test/legion/language/src/regent/std.t:1477: in function 'implicit_cast'
        ...maj/legion_s3d_test/legion/language/src/regent/codegen.t:2403: in function 'expr'
        ...maj/legion_s3d_test/legion/language/src/regent/codegen.t:9231: in function 'fn'
        src/terralist.lua:150: in function 'block'
        ...maj/legion_s3d_test/legion/language/src/regent/codegen.t:11172: in function 'compile_thunk'
        ...aj/legion_s3d_test/legion/language/src/regent/std_base.t:1632: in function 'complete'
        ...vsyamaj/legion_s3d_test/legion/language/src/regent/std.t:4518: in function <...vsyamaj/legion_s3d_test/legion/language/src/regent/std.t:4516>
        ...vsyamaj/legion_s3d_test/legion/language/src/regent/std.t:4520: in function 'compile_tasks_in_parallel'
        ...vsyamaj/legion_s3d_test/legion/language/src/regent/std.t:4654: in function 'start'
        test3.rg:144: in main chunk
...j/legion_s3d_test/legion/language/terra/src/terralib.lua:1202: Errors reported during typechecking.
/lustre/scratch/vsyamaj/legion_s3d_test/legion/language/src/regent/std.t:1477: invalid conversion from int1d to int3d
    return `([to]([expr]))
                      ^
/lustre/scratch/vsyamaj/legion_s3d_test/legion/language/src/regent/std.t:1477: user-defined cast failed: ...vsyamaj/legion_s3d_test/legion/language/src/regent/std.t:2385: assertion failed!
stack traceback:
        [C]: in function 'assert'
        ...vsyamaj/legion_s3d_test/legion/language/src/regent/std.t:2385: in function <...vsyamaj/legion_s3d_test/legion/language/src/regent/std.t:2360>
        [C]: in function 'xpcall'
        /lustre/scratch/vsyamaj/legion_s3d_test/legion/language/src/regent/std.t:1477: Errors reported during invoking __cast
    return `([to]([expr]))
                      ^
        ...j/legion_s3d_test/legion/language/terra/src/terralib.lua:2290: in function 'userfn'
        /lustre/scratch/vsyamaj/legion_s3d_test/legion/language/src/regent/std.t:1477: Errors reported during invoking macro
    return `([to]([expr]))
                 ^
        ...j/legion_s3d_test/legion/language/terra/src/terralib.lua:2859: in function 'docheck'
        ...j/legion_s3d_test/legion/language/terra/src/terralib.lua:3120: in function 'checkexp'
        ...j/legion_s3d_test/legion/language/terra/src/terralib.lua:3435: in function 'typecheck'
        ...j/legion_s3d_test/legion/language/terra/src/terralib.lua:1202: in function 'definequote'
        ...vsyamaj/legion_s3d_test/legion/language/src/regent/std.t:1477: in function 'implicit_cast'
        ...maj/legion_s3d_test/legion/language/src/regent/codegen.t:2403: in function 'expr'
        ...maj/legion_s3d_test/legion/language/src/regent/codegen.t:9231: in function 'fn'
        src/terralist.lua:150: in function 'block'
        ...maj/legion_s3d_test/legion/language/src/regent/codegen.t:11172: in function 'compile_thunk'
        ...aj/legion_s3d_test/legion/language/src/regent/std_base.t:1632: in function 'complete'
        ...vsyamaj/legion_s3d_test/legion/language/src/regent/std.t:4518: in function <...vsyamaj/legion_s3d_test/legion/language/src/regent/std.t:4516>
        ...vsyamaj/legion_s3d_test/legion/language/src/regent/std.t:4520: in function 'compile_tasks_in_parallel'
        ...vsyamaj/legion_s3d_test/legion/language/src/regent/std.t:4654: in function 'start'
        test3.rg:144: in main chunk
    return `([to]([expr]))
                      ^
/lustre/scratch/vsyamaj/legion_s3d_test/legion/language/src/regent/std.t:1477: user-defined cast failed: ...vsyamaj/legion_s3d_test/legion/language/src/regent/std.t:2385: assertion failed!
stack traceback:
        [C]: in function 'assert'
        ...vsyamaj/legion_s3d_test/legion/language/src/regent/std.t:2385: in function <...vsyamaj/legion_s3d_test/legion/language/src/regent/std.t:2360>
        [C]: in function 'xpcall'
        /lustre/scratch/vsyamaj/legion_s3d_test/legion/language/src/regent/std.t:1477: Errors reported during invoking __cast
    return `([to]([expr]))
                      ^
        ...j/legion_s3d_test/legion/language/terra/src/terralib.lua:2290: in function 'userfn'
        /lustre/scratch/vsyamaj/legion_s3d_test/legion/language/src/regent/std.t:1477: Errors reported during invoking macro
    return `([to]([expr]))
                 ^
        ...j/legion_s3d_test/legion/language/terra/src/terralib.lua:2859: in function 'docheck'
        ...j/legion_s3d_test/legion/language/terra/src/terralib.lua:3120: in function 'checkexp'
        ...j/legion_s3d_test/legion/language/terra/src/terralib.lua:3435: in function 'typecheck'
        ...j/legion_s3d_test/legion/language/terra/src/terralib.lua:1202: in function 'definequote'
        ...vsyamaj/legion_s3d_test/legion/language/src/regent/std.t:1477: in function 'implicit_cast'
        ...maj/legion_s3d_test/legion/language/src/regent/codegen.t:2403: in function 'expr'
        ...maj/legion_s3d_test/legion/language/src/regent/codegen.t:9231: in function 'fn'
        src/terralist.lua:150: in function 'block'
        ...maj/legion_s3d_test/legion/language/src/regent/codegen.t:11172: in function 'compile_thunk'
        ...aj/legion_s3d_test/legion/language/src/regent/std_base.t:1632: in function 'complete'
        ...vsyamaj/legion_s3d_test/legion/language/src/regent/std.t:4518: in function <...vsyamaj/legion_s3d_test/legion/language/src/regent/std.t:4516>
        ...vsyamaj/legion_s3d_test/legion/language/src/regent/std.t:4520: in function 'compile_tasks_in_parallel'
        ...vsyamaj/legion_s3d_test/legion/language/src/regent/std.t:4654: in function 'start'
        test3.rg:144: in main chunk
    return `([to]([expr]))
                      ^

stack traceback:
        [C]: in function 'error'
        ...j/legion_s3d_test/legion/language/terra/src/terralib.lua:396: in function 'finishandabortiferrors'
        ...j/legion_s3d_test/legion/language/terra/src/terralib.lua:3437: in function 'typecheck'
        ...j/legion_s3d_test/legion/language/terra/src/terralib.lua:1202: in function 'definequote'
        ...vsyamaj/legion_s3d_test/legion/language/src/regent/std.t:1477: in function 'implicit_cast'
        ...maj/legion_s3d_test/legion/language/src/regent/codegen.t:2403: in function 'expr'
        ...maj/legion_s3d_test/legion/language/src/regent/codegen.t:9231: in function 'fn'
        src/terralist.lua:150: in function 'block'
        ...maj/legion_s3d_test/legion/language/src/regent/codegen.t:11172: in function 'compile_thunk'
        ...aj/legion_s3d_test/legion/language/src/regent/std_base.t:1632: in function 'complete'
        ...vsyamaj/legion_s3d_test/legion/language/src/regent/std.t:4518: in function <...vsyamaj/legion_s3d_test/legion/language/src/regent/std.t:4516>
        ...vsyamaj/legion_s3d_test/legion/language/src/regent/std.t:4520: in function 'compile_tasks_in_parallel'
        ...vsyamaj/legion_s3d_test/legion/language/src/regent/std.t:4654: in function 'start'
        test3.rg:144: in main chunk
```

Run with MAX_DIM=4 and mpirun -np 2 regent.py test3.rg -ll:cpu 4

**Comment by elliottslaughter** [comment]
Should be fixed in daf80f5a2e77652ce425c4fa30ae3b393e168dd9.

**Comment by syamajala** [comment]
This works.
